### PR TITLE
Http client

### PR DESCRIPTION
**Pull request by dantleech**
Replaces #92 

@kelunik this also fixes https://github.com/dantleech/fink/issues/93

### CODE REVIEW

File: lib/Console/Display/ReportListDisplay.php [hunk 1]
```diff
@@ -18,8 +18,9 @@ public function render(OutputFormatterInterface $formatter, Status $status): str
             $statuses[] = sprintf(
                 $this->resolveFormat($index + 1 === count($status->reportStore()), $report),
                 sprintf(
-                    '[%3s] %s',
+                    '[%3s][%s] %s',
```
Comment (dantleech) [inline lib/Console/Display/ReportListDisplay.php:21]: Remove this (it will still be in the JSON)

File: tests/Example/custom-headers/headers [hunk 2]
```diff
@@ -0,0 +1,8 @@
+array (
+  'User-Agent' => 'Mozilla/5.0 (compatible; Artax; FinkPHP)',
+  'X-One' => 'Teapot',
+  'X-Two' => 'Pottea',
+  'accept' => '*/*',
+  'Accept-Encoding' => 'gzip, deflate, identity',
+  'host' => '127.0.0.1:8124',
+)
```
Comment (dantleech) [inline tests/Example/custom-headers/headers:8]: remove this

File: lib/Model/Crawler.php [hunk 3]
```diff
@@ -26,43 +27,45 @@ public function crawl(Url $documentUrl, UrlQueue $queue, ReportBuilder $report):
     {
         $start = microtime(true);
         $report->withReferringElement($documentUrl->referringElement());
-        $response = yield $this->client->request($documentUrl->__toString());
+        $response = yield $this->client->request(new Request((string) $documentUrl));
         $time = (microtime(true) - $start) * 1E6;
+        assert($response instanceof Response);
+
+        // if there was a redirect, use the redirect URL as the base
+        $ultimateUrl = $documentUrl->withPsiUrl($response->getRequest()->getUri());
```
Comment (dantleech) [inline lib/Model/Crawler.php:35]: this was the fix for the URL redirection issue

File: lib/Model/Url.php [hunk 4]
```diff
@@ -42,49 +45,45 @@ private function __construct(Uri $uri, Url $referrer = null, int $distance = 0,
...
-        } catch (Exception $e) {
+            $new = new self(HttpUri::createFromString($url));
+        } catch (\Throwable $e) {
             throw new InvalidUrl($e->getMessage(), 0, $e);
         }
 
         return $new;
     }
 
+    public function withPsiUrl(PsrUri $uri): self
```
Comment (kelunik) [inline lib/Model/Url.php:56]: withPsrUrl?

Comment (dantleech) [reply lib/Model/Url.php:56]: s/Url/Uri

**Comment by dantleech** [comment]
:boom:

**Pull request merged**